plugins/snowflake/dbt/adapters/snowflake/connections.py

```diff
@@ -47,10 +70,63 @@ def auth_args(self):
         if self.authenticator:
             result['authenticator'] = self.authenticator
             if self.authenticator == 'oauth':
-                result['token'] = self.token
+                token = self.token
+                # if we have a client ID/client secret, the token is a refresh
+                # token, not an access token
+                if self.oauth_client_id and self.oauth_client_secret:
+                    token = self._get_access_token()
+                elif self.oauth_client_id:
+                    warn_or_error(
+                        'Invalid profile: got an oauth_client_id, but not an '
+                        'oauth_client_secret!'
+                    )
+                elif self.oauth_client_secret:
+                    warn_or_error(
+                        'Invalid profile: got an oauth_client_secret, but not '
+                        'an oauth_client_id!'
+                    )
+
+                result['token'] = token
         result['private_key'] = self._get_private_key()
         return result
 
+    def _get_access_token(self) -> str:
+        if self.authenticator != 'oauth':
+            raise InternalException('Can only get access tokens for oauth')
+        missing = any(
+            x is None for x in
+            (self.oauth_client_id, self.oauth_client_secret, self.token)
+        )
+        if missing:
+            raise InternalException(
+                'need a client ID a client secret, and a refresh token to get '
+                'an access token'
+            )
+        # should the full url be a config item?
+        token_url = _TOKEN_REQUEST_URL.format(self.account)
+        # I think this is only used to redirect on success, which we ignore
+        # (it does not have to match the integration's settings in snowflake)
+        redirect_uri = 'http://localhost:9999'
+        data = {
+            'grant_type': 'refresh_token',
+            'refresh_token': self.token,
+            'redirect_uri': redirect_uri
+        }
+
+        auth = base64.b64encode(
+            f'{self.oauth_client_id}:{self.oauth_client_secret}'
+            .encode('ascii')
+        ).decode('ascii')
+        headers = {
+            'Authorization': f'Basic {auth}',
+            'Content-type': 'application/x-www-form-urlencoded;charset=utf-8'
+        }
+        result = requests.post(token_url, headers=headers, data=data)
+        result_json = result.json()
+        if 'access_token' not in result_json:
+            raise DatabaseException(f'Did not get a token: {result_json}')
```

Comment:
do you know what kind of contents would be returned in `result_json` here? I'm just worried that this could print sensitive-ish info out to the console, like a client id or client secret. If there's a chance that could happen, we should probably stringify the json and mask out creds if we can!

Reply:
I usually got error messages about invalid client IDs. I don't think it returns sensitive information, it was actually very annoying to debug because of that (which is good/normal security practice for an endpoint like this):
```
Encountered an error:
Database Error
  Did not get a token: {'data': None, 'message': 'This is an invalid client.', 'code': None, 'success': False, 'error': 'invalid_client'}
```